lib/climate_control.rb

```diff
@@ -1,16 +1,38 @@
-require "climate_control/environment"
 require "climate_control/errors"
-require "climate_control/modifier"
 require "climate_control/version"
+require "monitor"
 
 module ClimateControl
-  @@env = ClimateControl::Environment.new
+  extend self
 
-  def self.modify(environment_overrides, &block)
-    Modifier.new(env, environment_overrides, &block).process
+  SEMAPHORE = Monitor.new
+  private_constant :SEMAPHORE
+
+  def modify(environment_overrides, &block)
+    environment_overrides = environment_overrides.transform_keys(&:to_s)
+
+    SEMAPHORE.synchronize do
+      pre = mid = env.to_hash
+      copy environment_overrides, into: env
+      mid = env.to_hash
+      block.call
+    ensure
+      (pre.keys | mid.keys | env.keys).each do |key|
+        env[key] = pre[key] if pre[key] != env[key] && mid[key] == env[key]
+      end
```

Comment:
I think this part is quite hard to understand especially with the indirect `env/ENV` access and the names of the variables

Reply:
IIIRC, my reasoning there was that it could be overridden for testing purposes. Eg it would be nice to be able to test the code without it messing up the env variables and potentially affecting the test suite. Though in practice it may not be that simple, since `ENV` isn't totally equivalent to a hash (eg assigning a value of `nil` to a key is the same as deleting it for `ENV`, but not for a hash).